TripalCultivate/TripalCultivate-Phenotypes
# G4.92 Update validator base, annotation and interface

**Issue #92**

## Motivation

Issue https://github.com/TripalCultivate/TripalCultivate-Phenotypes/issues/82 describes changes to the design of the validator api. This issue brings those changes into the annotation, interface and base validator class. It also marks the old design components as deprecated.

## What does this PR do?

1. Moves the validator plugin base into a dedicated directory and update namespace and class references.
2. Adds validateMetadata(), validateFile() and validateRow() to interface and base class.
3. Documents new return value and old one.
4. Adds inputTypes to annotation.
5. Adds getSupportedInputTypes() and checkInputTypeSupported() to support inputType usage.
6. Mark all deprecated functions and parameters with `@deprecated`

## Testing

### Automated Testing

In the ValidatorBaseTest all of my new test methods create a validator developed specifically for testing the base class. This validator is found in `trpcultivate_phenotypes/tests/src/Kernel/Validators/FakeValidators/BasicallyBase.php` and simply extends the base class without implementing or overriding anything. As such, it is essentially the base class with added annotation 😜 

ValidatorBaseTest->testBasicValidatorGetters: tests retrieving the name and the configuration allow_new which indicates if new traits can be added.
- retrieves the name of the BasicallyBase validator using getValidatorName()and compares it to what I wrote in the annotation block.
- retrieves the status of the allow_new configuration and confirms it matches the default which is TRUE. Note that this is independent from the importer but already existed in the base class.

ValidatorBaseTest->testInputTypeValidatorGetters: tests the two new methods I added related to the inputType in the annotation. There's an example of the inputType annotation being used in the BasicallyBase and ValueInList validators.
- retrieves the value of the inputType in the BasicallyBase validator using getSupportedInputTypes(). This should return an array and we confirm that this array matches the one I wrote in the annotation. I use assertEqualsCanonicalizing to ensure that the order of the array does not cause the tests to fail.
- asks the BasicallyBase validator if it supports the `data-row` inputType and confirms that it knows it does.
- asks the BasicallyBase validator if it supports the `metadata` inputType and confirms that it knows it does NOT.
- asks the BasicallyBase validator if it supports the non-existant `SARAH` inputType and confirms that it knows it does NOT.

ValidatorBaseTest->testValidatorValidateMethods: tests the base implementation of the validate* methods. The base implementation is meant to simply inform the developer that the need to override it to support it. They're also there so that they can all be defined in the interface for consistency without requiring all validators to implement all of them.
- calls validateMetadata with random params matching the type hints and confirms we get an exception specific to validateMetadata.
- calls validateFile with random params matching the type hints and confirms we get an exception specific to validateFile.
- calls validateRow with random params matching the type hints and confirms we get an exception specific to validateRow.
- calls validate with random params matching the type hints and confirms we get an exception specific to validate. Note that the test for this method will change to check for the deprecation method once we get near the end of this design update.

#### Test Coverage

<img width="1222" alt="Screenshot 2024-06-28 at 9 55 30 AM" src="https://github.com/TripalCultivate/TripalCultivate-Phenotypes/assets/1566301/849f34d3-63f1-4ca0-b769-5c518a26bd4e">

<img width="1220" alt="Screenshot 2024-06-28 at 9 55 01 AM" src="https://github.com/TripalCultivate/TripalCultivate-Phenotypes/assets/1566301/f20701ea-19d3-43f8-8399-25516cc39f7b">

### Manual Testing

As this changes the interface, annotation and abstract base class it's quite hard to test 🤔 

Unfortunately the BasicallyBase plugin is not accessible in the drush php:cli so we can't just use the same approach as in the automated tests. In order to test cases I didn't or different values you will need to edit the automated tests. 

**Thankfully, we can test the ValuesInList validator in drush php:cli and I've updated it so it can be used as an example.**

For the annotation and such you can also test the ValuesInList validator as I updated it to show an example of the inputType. 

```php
$plugin_manager = \Drupal::service('plugin.manager.trpcultivate_validator');
$validator_id = 'trpcultivate_phenotypes_validator_value_in_list';
$instance = $plugin_manager->createInstance($validator_id);

$instance->getValidatorName();
$instance->getSupportedInputTypes();
$instance->checkInputTypeSupported('data-row');
$instance->checkInputTypeSupported('metadata');
$instance->checkInputTypeSupported('SARAH');
```
<img width="820" alt="Screenshot 2024-06-28 at 10 35 28 AM" src="https://github.com/TripalCultivate/TripalCultivate-Phenotypes/assets/1566301/9c23f35a-9fd0-478e-a954-c3c36440ee43">

Testing the validate functions you expect to see the exception for all but the validateValues() which this validator does override.

```php
$plugin_manager = \Drupal::service('plugin.manager.trpcultivate_validator');
$validator_id = 'trpcultivate_phenotypes_validator_value_in_list';
$instance = $plugin_manager->createInstance($validator_id);

$form_values = ['genus' => 'Fred', 'project_id' => 123];
$instance->validateMetadata($form_values);

$filename = 'public://does_not_exist.txt';
$fid = 123;
$instance->validateFile($filename, $fid);

$row_values = ['col1', 'col2', 'col3', 'col4', 'col5'];
$context = [
  'indices' => [0, 2], 
  'valid_values' => ['col3']
];
$instance->validateRow($row_values, $context);

$instance->validate();
```
<img width="1583" alt="Screenshot 2024-06-28 at 10 42 33 AM" src="https://github.com/TripalCultivate/TripalCultivate-Phenotypes/assets/1566301/1244d7e8-7fb6-4a9c-9c45-0c5e957054c4">
